spec_validator.go

```diff
@@ -0,0 +1,148 @@
+	eventPrefix  = "event"
+	dnsPrefix    = "dns"
+)
+
+var (
+	ErrorInvalidMessageEncoding = errors.New("invalid message encoding")
+	ErrorInvalidMessageType     = errors.New("invalid message type")
+	ErrorInvalidSource          = errors.New("invalid Source name")
+	ErrorInvalidDestination     = errors.New("invalid Destination name")
+)
+
+// SpecValidators is a WRP validator that ensures messages are valid based on
+// each spec validator in the list. Only validates the opinionated portions of the spec.
+// SpecValidators validates the following fields: UTF8 (all string fields), MessageType, Source, Destination
+func SpecValidators() Validators {
+	return Validators{UTF8Validator, MessageTypeValidator, SourceValidator, DestinationValidator}
+}
+
+// UTF8Validator is a WRP validator that takes messages and validates that it contains UTF-8 strings.
+var UTF8Validator ValidatorFunc = func(m Message) error {
+	if err := UTF8(m); err != nil {
+		return fmt.Errorf("%w: %v", ErrorInvalidMessageEncoding, err)
+	}
+
+	return nil
+}
+
+// MessageTypeValidator is a WRP validator that takes messages and validates their Type.
+var MessageTypeValidator ValidatorFunc = func(m Message) error {
+	t := m.MessageType()
+	if t < Invalid0MessageType || t > lastMessageType {
+		return ErrorInvalidMessageType
+	}
+
+	switch t {
+	case Invalid0MessageType, Invalid1MessageType, lastMessageType:
+		return ErrorInvalidMessageType
+	}
+
+	return nil
+}
+
+// SourceValidator is a WRP validator that takes messages and validates their Source.
+// Only mac and uuid sources are validated. Serial, event and dns sources are
+// not validated.
+var SourceValidator ValidatorFunc = func(m Message) error {
+	if err := validateLocator(m.Source); err != nil {
+		return fmt.Errorf("%w: %v", ErrorInvalidSource, err)
+	}
+
+	return nil
+}
+
+// DestinationValidator is a WRP validator that takes messages and validates their Destination.
+// Only mac and uuid destinations are validated. Serial, event and dns destinations are
+// not validated.
+var DestinationValidator ValidatorFunc = func(m Message) error {
+	if err := validateLocator(m.Destination); err != nil {
+		return fmt.Errorf("%w: %v", ErrorInvalidDestination, err)
+	}
+
+	return nil
+}
+
+// locatorPattern is the precompiled regular expression that all source and dest locators must match.
+// Matching is partial, as everything after the authority (ID) is ignored. https://xmidt.io/docs/wrp/basics/#locators
+var locatorPattern = regexp.MustCompile(
+	`^(?P<scheme>(?i)` + macPrefix + `|` + uuidPrefix + `|` + eventPrefix + `|` + dnsPrefix + `|` + serialPrefix + `):(?P<authority>[^/]+)?`,
+)
+
+// validateLocator validates a given locator's scheme and authority (ID).
+// Only mac and uuid schemes' IDs are validated. IDs from serial, event and dns schemes are
+// not validated.
+func validateLocator(l string) error {
+	match := locatorPattern.FindStringSubmatch(l)
+	if match == nil {
+		return fmt.Errorf("spec scheme not found")
```

Comment:
spec scheme not found is a bit confusing (at least for me) - what about something like `value given doesn't match expected locator pattern`?  Then the error would be wrapped as...
```
invalid Source name 'abcd': value given doesn't match expected locator pattern
```